core/src/main/java/com/datastrato/gravitino/storage/relational/service/FilesetMetaService.java

```diff
@@ -0,0 +1,240 @@
+                      mapper.insertFilesetMetaOnDuplicateKeyUpdate(po);
+                    } else {
+                      mapper.insertFilesetMeta(po);
+                    }
+                  }),
+          () ->
+              SessionUtils.doWithoutCommit(
+                  FilesetVersionMapper.class,
+                  mapper -> {
+                    if (overwrite) {
+                      mapper.insertFilesetVersionOnDuplicateKeyUpdate(po.getFilesetVersionPO());
+                    } else {
+                      mapper.insertFilesetVersion(po.getFilesetVersionPO());
+                    }
+                  }));
+    } catch (RuntimeException re) {
+      ExceptionUtils.checkSQLConstraintException(
+          re, Entity.EntityType.FILESET, filesetEntity.nameIdentifier().toString());
+      throw re;
+    }
+  }
+
+  public <E extends Entity & HasIdentifier> FilesetEntity updateFileset(
+      NameIdentifier identifier, Function<E, E> updater) throws IOException {
+    NameIdentifier.checkFileset(identifier);
+
+    String filesetName = identifier.name();
+
+    Long schemaId =
+        CommonMetaService.getInstance().getParentEntityIdByNamespace(identifier.namespace());
+
+    FilesetPO oldFilesetPO = getFilesetPOBySchemaIdAndName(schemaId, filesetName);
+    FilesetEntity oldFilesetEntity =
+        POConverters.fromFilesetPO(oldFilesetPO, identifier.namespace());
+    FilesetEntity newEntity = (FilesetEntity) updater.apply((E) oldFilesetEntity);
+    Preconditions.checkArgument(
+        Objects.equals(oldFilesetEntity.id(), newEntity.id()),
+        "The updated fileset entity id: %s should be same with the table entity id before: %s",
+        newEntity.id(),
+        oldFilesetEntity.id());
+
+    Integer updateResult;
+    try {
+      boolean checkNeedUpdateVersion =
+          POConverters.checkFilesetVersionNeedUpdate(oldFilesetPO.getFilesetVersionPO(), newEntity);
+      FilesetPO newFilesetPO =
+          POConverters.updateFilesetPOWithVersion(oldFilesetPO, newEntity, checkNeedUpdateVersion);
+      if (checkNeedUpdateVersion) {
+        // These operations are guaranteed to be atomic by the transaction. If version info is
+        // inserted successfully and the uniqueness is guaranteed by `fileset_id + version +
+        // deleted_at`, it means that no other transaction has been inserted (if a uniqueness
+        // conflict occurs, the transaction will be rolled back), then we can consider that the
+        // fileset meta update is successful
+        SessionUtils.doMultipleWithCommit(
+            () ->
+                SessionUtils.doWithoutCommit(
+                    FilesetVersionMapper.class,
+                    mapper -> mapper.insertFilesetVersion(newFilesetPO.getFilesetVersionPO())),
+            () ->
+                SessionUtils.doWithoutCommit(
+                    FilesetMetaMapper.class,
+                    mapper -> mapper.updateFilesetMeta(newFilesetPO, oldFilesetPO)));
```

Comment:
For update sql, should we check the return value. Because if there is no item which we want to update, the update result will return 0 instead of a thrown exception.

Reply:
One request deleted the fileset. Another request update the fileset. The fileset is absent, the update operation will be executed successfully with update result 0.

Reply:
Yes, this problem will occur if the upper layer does not have a lock manager's read-write lock. However, according to the current implementation, our server is a single-node service, and there may be no need to consider this concurrency scenario for the time being. I need to see how we consider the multi-node situation in the future and whether we need to implement read-write locks at the storage layer.

Reply:
OK.